iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/exchange/DefaultTokenExchangePdp.java

```diff
@@ -99,14 +109,29 @@ private TokenExchangePdpResult verifyScopes(TokenExchangePolicy p, TokenRequest
       return fromPolicy(p);
     }
 
-    // The requested scopes must be allowed by the origin client (destination is impersonating the
-    // origin client)
-    Set<ScopeMatcher> originClientMatchers = scopeMatcherRegistry.findMatchersForClient(origin);
+    Set<ScopeMatcher> scopeMatchers = scopeMatcherRegistry.findMatchersForClient(origin);
+    String invalidScopeMessage = "scope not allowed by origin client configuration";
+    String subjectToken = request.getRequestParameters().get("subject_token");
+    if (properties.getJwtProfile().isTokenExchangeDisableUpscoping() && !subjectToken.isBlank()) {
+      invalidScopeMessage = "scope not allowed by subject token configuration";
+      try {
+        // this assumes that iam.access_token.include_scope=true
```

Comment:
in case it's false probably scopes can be retrieved from TokenRequest object. 
We could try to implement a method "extractScopes" that receives the subjectToken and the request as arguments and returns the String[] of scopes

Reply:
Catching Throwable instead of Exception could be a solution. Also checking the incòude_scope value too.